Raise feature not supported error for SELECT FOR JSON for various datatypes

anujsarda3 opened the pull request:
### Description

- Raising feature not supported error for varbinary, binary, rowversion, image, timestamp, datetime2, datetime, smalldatetime, datetimeoffset, bit, money and smallmoney datatypes
- Added test cases to check each and every datatype
- Moving enum declaration for FOR JSON from engine's parser.h to extensions forjson.h
- This is not a complete fix for the issue, it raises featured not supported error for unsupported datatypes
 
Engine PR - https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/pull/62

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

### Issues Resolved

BABEL-3691


By submitting this pull request, I confirm that my contribution is under the terms of the Apache 2.0 and PostgreSQL licenses, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/babelfish_extensions/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

anujsarda3 committed:
SELECT FOR JSON support for various datatypes

- Added support for various dataypes and raised error for the datatypes that are not
  supported
- Added test cases to check each and every datatype

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Added test cases to latest/schedule

- Added test case to the latest/schedule file

TASK: BABEL-3691
Signed-off-by: Anuj Sarda<sardanuj@amazon.com>

Sairakan commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -55,6 +57,31 @@ SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_val
+			strcmp(typename, "smallmoney") == 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("data type exact numeric's money or smallmoney is not supported")));
```
message wording nitpick: leave it as "data types money or smallmoney are not supported with FOR JSON"

2jungkook commented on test/JDBC/input/BABEL-3691-vu-verify.sql:
```diff
@@ -0,0 +1,37 @@
+-- DIFFERENT CASES TO CHECK DATATYPES
+-- EXACT NUMERICS
+select a, b, c, d, f, g, i from dt01 for json path
```
Please use `proc` and `view` instead of a raw `select`.

2jungkook commented on test/JDBC/input/BABEL-3691-vu-prepare.sql:
```diff
@@ -0,0 +1,41 @@
+-- DIFFERENT CASES TO CHECK DATATYPES
+-- EXACT NUMERICS
+create table dt01(a bigint, b bit, c decimal, d int, e money, f numeric, g smallint, h smallmoney, i tinyint)
```
Our convention for object name is something like `BABEL_3691_vu_prepare_t1`.

2jungkook commented on test/JDBC/input/BABEL-3691-vu-verify.sql:
```diff
@@ -0,0 +1,37 @@
+select a, b, c, d, f, g, i from dt01 for json path
+go
+
+select e from dt01 for json path
```
Please add test cases regarding explicit conversions, e.g.,
```
select col1=cast('2022-11-11 20:56:22.41' as datetime) from dt01 for json path
```

anujsarda3 replied:
resolving this, as currently we are only raising error for datatypes. will make to add when feature will be supported.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -55,6 +57,31 @@ SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_val
+	for (i = 1; i <= SPI_tuptable->tupdesc->natts; i++)
+	{
+		char* typename = SPI_gettype(SPI_tuptable->tupdesc, i);
+		if (strcmp(typename, "binary") == 0 ||
```
Please check the namespace `sys` as well.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -55,6 +57,31 @@ SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_val
+		if (strcmp(typename, "binary") == 0 ||
+			strcmp(typename, "varbinary") == 0 ||
+			strcmp(typename, "image") == 0 ||
+			strcmp(typename, "timestamp") == 0)
```
What about `rowversion` type?

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -55,6 +57,31 @@ SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_val
+			strcmp(typename, "timestamp") == 0)
+			ereport(ERROR,
+					(errcode(ERRCODE_FEATURE_NOT_SUPPORTED),
+						errmsg("option binary base64 is not supported")));
```
1. Is this a proper error message for `image`, `timestamp` and `rowversion`?
2. Please add test cases for all unsupported data types.

anujsarda3 replied:
Adding [link](https://learn.microsoft.com/en-us/sql/relational-databases/json/how-for-json-converts-sql-server-data-types-to-json-data-types-sql-server?view=sql-server-ver16#:~:text=The%20FOR%20JSON%20clause%20uses%20the%20following%20rules%20to%20convert%20SQL%20Server%20data%20types%20to%20JSON%20types%20in%20the%20JSON%20output) for reference.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -55,6 +57,31 @@ SPI_sql_row_to_json_path(uint64 rownum, StringInfo result, bool include_null_val
+	* For now report an ERROR if any attribute is binary data type since base64
+	* encoding is not implemented yet.
+	* Also Exact numeric's datatype money or smallmoney is not supported.
+	*/
```
```
/*
 *
 */
```

2jungkook commented:
Please add the followings in the PR description
1. Engine PR's link (do the same for the Engine PR's description)
2. list of supported and unsupported data types

anujsarda3 committed:
Feature not supported for unsupported datatypes

- Raising feature not supported error for all unsupported datatypes
- Added enum declaration in gram-tsql-prologue.y.h

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3691

anujsarda3 committed:
Nit edits in forjson.c

- Nit edits in forjson.c

TASK: BABEL-3691
Signed-off-by: Anuj Sarda<sardanuj@amazon.com>

anujsarda3 committed:
Added test cases for rowversion and timestamp

- Added test case for rowversion and timestamp

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3691

anujsarda3 committed:
Implementation of tsql datatype oid lookup

- Implemented the function to lookup tsql datatype oid

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Comment changes

- change in comment in function tsql_unsupported_datatype_check

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3691

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -101,7 +107,7 @@ tsql_query_to_json_internal(const char *query, int mode, bool include_null_value
 	StringInfo	result;
 	uint64		i;
-
+	
```
Please remove this.

anujsarda3 replied:
Although this was added keeping in mind the format of other functions in the same file, having a space after variable declarations, will remove the space in my next commit.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -112,6 +118,15 @@ tsql_query_to_json_internal(const char *query, int mode, bool include_null_value
+	}
+
+	// To check if query output table has columns with datatypes that are currently not supported in FOR JSON
+	tsql_unsupported_datatype_check();
```
We should catch the exception and call `SPI_finish()`.

anujsarda3 replied:
Resolving as discussed.

2jungkook commented on contrib/babelfishpg_tsql/src/forjson.c:
```diff
@@ -145,6 +160,64 @@ tsql_query_to_json_internal(const char *query, int mode, bool include_null_value
+
+		tsql_datatype_oid = GetSysCacheOid2(TYPENAMENSP, Anum_pg_type_oid, CStringGetDatum(typename), ObjectIdGetDatum(nspoid));
+
+		if (tsql_datatype_oid == datatype_oid)
```
Please add a comment explaining when `tsql_datatype_oid` could be different from `datatype_oid`.

2jungkook commented on test/JDBC/expected/BABEL-3691-vu-cleanup.out:
```diff
@@ -0,0 +1,97 @@
+GO
+
+-- Rowversion and timestamp
+EXEC sp_babelfish_configure 'babelfishpg_tsql.escape_hatch_rowversion', 'ignore';
```
Is this necessary?

anujsarda3 replied:
This was added with reference from other SQL test files. But yes, this is not necessary. Will remove it.

2jungkook commented on test/JDBC/expected/BABEL-3691-vu-verify.out:
```diff
@@ -0,0 +1,135 @@
+
+
+-- Rowversion and timestamp
+EXEC sp_babelfish_configure 'babelfishpg_tsql.escape_hatch_rowversion', 'ignore';
```
Do we need to change the escape hatch even before `select`?

anujsarda3 committed:
Changes in test files and added comment

- Remove EXEC command from test files
- Added comments as needed in forjson.c

TASK: BABEL-3691
Signed-off-by: Anuj Sarda <sardanuj@amazon.com>

anujsarda3 committed:
Merge remote-tracking branch 'upstream/BABEL_2_X_DEV' into jira-babel-3691

anujsarda3 committed:
Created Header file forjson.h

- created header file forjson.h and done enum declaration instead of
  directly including gram-tsql-prologue.y.h file in forjson.c

TASK: BABEL-3691
Signed-off-by: Anuj Sarda<sardanuj@amazon.com>
